Adding tutorials in benchopt doc

cohenjer opened the pull request:
This PR proposes to add Tutorials and Examples to the documentation of Benchopt. At the core of this PR is the [DIVIO](https://documentation.divio.com/) documentation concept. The documentation of Benchopt at the moment is heavily oriented towards API and mechanical functioning (how things work). It may be lacking in Tutorials and Examples contents (how to use Benchopt).

At the moment I have only worked on Tutorials.

Things added:
---------------------
- Adding an example and tutorial tab
- Adding a tutorial on how to write a benchmark

A few things todo as well (coming soon):
-----------------------------------------------------------
- Update the template benchmark to unify notations (beta and w)
- Reorganise the images used in the benchmark, I put them in the same folder as the .rst file but there should be a better practice.

Any feedback on the tutorial itself, devoted to writing a first Benchmark, is welcome!

agramfort commented:
here is the rendered doc if you want to check it out https://output.circle-artifacts.com/output/job/5bbea0ef-cfcf-415c-8e67-a227d61d4329/artifacts/0/dev/index.html

agramfort commented:
@cohenjer you need to update https://github.com/benchopt/benchopt/blob/main/MANIFEST.in to fix the CI

cohenjer commented:
> @cohenjer you need to update https://github.com/benchopt/benchopt/blob/main/MANIFEST.in to fix the CI

Done, I added *.svg in the include

agramfort reviewed (approved):
provided CIs are happy

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+.. _Benchopt: https://benchopt.github.io/#
+
+Understanding and building a simple benchmark with Benchopt
```
is "with Benchopt" necessary here?

cohenjer replied:
I guess not, removing it!

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+Understanding and building a simple benchmark with Benchopt
+===========================================================
+
+Are you looking for a simple, step-by-step introduction to the `Benchopt`_ library? Then look no further! In this tutorial, we are going to learn how to use the `Benchopt`_ library to compare various algorithms in order to minimize a given cost. This tutorial complements the `write a benchmark <https://benchopt.github.io/how.html>`_ webpage with a more practical, hands-on approach. 
```
IMO no need to make Benchopt a link (2 occurrences)

cohenjer replied:
I removed the global link definition. I think it is still nice if Benchopt homepage is linked at least once, in case this html is used outside the doc for some reason.

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+Are you looking for a simple, step-by-step introduction to the `Benchopt`_ library? Then look no further! In this tutorial, we are going to learn how to use the `Benchopt`_ library to compare various algorithms in order to minimize a given cost. This tutorial complements the `write a benchmark <https://benchopt.github.io/how.html>`_ webpage with a more practical, hands-on approach. 
+
+Let's say we want to minimize the cost function of the `ridge regression problem <https://en.wikipedia.org/wiki/Ridge_regression>`_, a variant of linear regression which is a common of supervised machine learning.
```
Maybe: Lets' say you're interested in benchmarking solvers to the...

mathurinm replied:
typo "which is a common of supervised learning"

cohenjer replied:
tried something!

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+.. image:: screenshot_mytemplate_ridgeloss.png
+    :width: 800
+
+We can now dive into the framework and see how leverage it to easily create a benchmark!
```
```suggestion
We can now dive into the framework and see how to leverage it to easily create our benchmark!
```

cohenjer replied:
done

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+Installing Benchopt
+-------------------
+
+Let's start by installing the benchopt package. It is a python package, so we are assuming that you are starting this tutorial with a functioning python installation, preferably version >=3.10. It is recommanded to run all this tutorial in a dedicated environment (such as a `virtual environment <https://docs.python.org/fr/3/library/venv.html>`_ or a conda environment)
```
I'd also remove the mention of virtual environments and conda environments (orthgonal to the tuto IMO). YMMV

cohenjer replied:
removed the python version bit.

If possible I would like to keep the mention of virtual environment (at least until more reviews suggest we remove it), it is a good practice anyways and will avoid that people have issues with the tutorial because of a cluttered (base) environment.

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+Benchopt and its dependencies are going to be installed, just wait until this is done. If you have any issue with the installation, if you are already in a clean virtual environment, you may have a look at the `installation page <https://benchopt.github.io/#install>`_.
+
+Installing and Running a first benchmark
```
```suggestion
Installing and running a first benchmark
```

cohenjer replied:
done

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+Installing and Running a first benchmark
+----------------------------------------
+
+Benchopt in itself is like a mother package: it supervises and runs smaller libraries, a.k.a. the benchmarks.Therefore, benchmarks are repositories with a predefined structure and some required files (we will go over each of them in the next section) that are processed by benchopt. The benchmark specifies information about the problem (typically the loss function to minimize), the dataset on whic ...
```
```suggestion
Benchopt in itself is like a mother package: it supervises and runs smaller libraries, a.k.a. the benchmarks. Therefore, benchmarks are repositories with a predefined structure and some required files (we will go over each of them in the next section) that are processed by benchopt. The benchmark specifies information about the problem (typically the loss function to minimize), the dataset on which the algorithms are tested (values of \\(X \\) and \\( y\\)) and the solvers (like our GD algorithm). It is likely that if you are reading this tutorial, you are in fact mostly interested in writing and running a benchmark.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+Benchopt in itself is like a mother package: it supervises and runs smaller libraries, a.k.a. the benchmarks.Therefore, benchmarks are repositories with a predefined structure and some required files (we will go over each of them in the next section) that are processed by benchopt. The benchmark specifies information about the problem (typically the loss function to minimize), the dataset on whic ...
+
+To write a benchmark, it is recommended to start from the template benchmark shared in the Benchopt organisation. To get this template and rename it ``my_benchmark``, you can clone it from its Github repository
```
add a link here

cohenjer replied:
done

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+    g(\beta) = \frac{1}{2} \|y - X\beta \|^2
+
+and the solver implement is GD with \\(\\nabla_g(\\beta) = -X^Ty + X^TX\\beta \\) the gradient of \\(g\\) at \\(\\beta \\).
```
```suggestion
and the solver implement is GD with \\(\\nabla g(\\beta) = -X^Ty + X^TX\\beta \\) the gradient of \\(g\\) at \\(\\beta \\).
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+and the solver implement is GD with \\(\\nabla_g(\\beta) = -X^Ty + X^TX\\beta \\) the gradient of \\(g\\) at \\(\\beta \\).
+
+We will modify this template to adapt it to the ridge regression problem next, but before that let us run this benchmark. In other words, let us use benchopt to read the contents of the template benchmark and run GD on OLS with a predefined set of stepsizes, in our case \\( [1, 1.99] \\) (the stepsize is scaled with the Lipschitz constant, you can ignore this detail if you are not familiar with t ...
```
```suggestion
We will modify this template to adapt it to the ridge regression problem next, but before that let us run this benchmark. In other words, let us use benchopt to read the contents of the template benchmark and run GD on OLS with a predefined set of stepsizes, in our case \\( [1, 1.99] \\) (the stepsize is scaled by the inverse of the gradient's Lipschitz constant - you can ignore this detail if you are not familiar with this concept).
```

cohenjer replied:
thanks I was being too sketchy ^^

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+    - *Ordinary Least Squares* tells us which loss is minimized, and the hyperparameters are written in bracket. 
+    - *GD* is a line indicating the progress of algorithm GD for this problem (Simulated dataset, OLS loss). Again its hyperparameters are written in brackets (here the stepsize value).
+
+Once the benchmark has been run, a pop-up window should open in your default navigator. This provides a visualization of the results of the run, which is useful to immediately observe, comment and share the results. After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset. The convergence plots ca ...
```
```suggestion
Once the benchmark has been run, a window should open in your default navigator. This provides a visualization of the results of the run, which is useful to immediately observe, comment and share the results. After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset. The convergence plots can be made log-log for easier reading. Also, depending on your computer, the runtime is so low that the convergence plot with respect to time is not reliable. We advise to look at the results in terms of iteration (scrolling menu on the bottom left of the webpage). Feel free to play around with the plotting options here! Note that the dataset comes with two values for an option whiten, let us ignore this detail in the tutorial.
```

cohenjer replied:
done

mathurinm reviewed (changes requested):
Thanks for the great pass, I did a first pass with minor changes.
 
@cohenjer when versioning blocks of text I find it makes my life easier to break line at the end of every sentence (it does not change the rendering, but makes the diffs more compact). Could you please do it?

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+.. math:: 
+
+    \beta \leftarrow \beta - \eta \nabla_{f}(\beta)
```
done

mathurinm replied:
btw you can commit suggestions directly from the github interface (and commit them in batch to avoid having 1 commit per suggested modif)

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+where :math:`\eta` is the learning rate or stepsize, and controls how much the unknown is moved alongside the descent direction. Vector :math:`\nabla_{f}(\beta)` is the gradient of the cost function defined above, evaluated at \\(\\beta\\).
+
+Even though GD is a simple algorithm, when applied to our ridge regression problem, we should be mindful of the two hyperparameters: the regularization parameter \\(\\lambda\\), and the gradient stepsize \\( \\eta\\). It is very common in machine learning that we are facing a situation where we want to vary these parameters, and compare the behavior of our algorithm in each scenario. Benchopt we  ...
```
:)

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+Even though GD is a simple algorithm, when applied to our ridge regression problem, we should be mindful of the two hyperparameters: the regularization parameter \\(\\lambda\\), and the gradient stepsize \\( \\eta\\). It is very common in machine learning that we are facing a situation where we want to vary these parameters, and compare the behavior of our algorithm in each scenario. Benchopt we  ...
+
+.. image:: screenshot_mytemplate_ridgeloss.png
```
GD and GD-ridge are the two solvers we get at the end of the tutorial. As stated, this is a screenshot from the future, I think it is fine if not everything is clear yet, since explaining this is the whole point of the tutorial.

The name of the Objective (OLS) is something I forgot to change in the tutorial.

By naming the benchmark my_benchmark, the plot prints My Benchmark benchmark as a title, which I agree is not great. Any suggestion?

tomMoral replied:
The `GD` solver is designed for OLS and we change the benchmark to tackle Ridge regression.
And we also want the tuto to explain how to add a solver. So the idea was to had a correct solver, vs the existing one that does not account for regularisation. This allow to show an "interesting" thing, that is that if you don't optimize the right problem, it does not work :)

But maybe we can simply modify the GD one directly and add a diverging one with step size too large? I have no strong opinion here.

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+    \beta \leftarrow \beta - \eta \nabla_{f}(\beta)
+
+where :math:`\eta` is the learning rate or stepsize, and controls how much the unknown is moved alongside the descent direction. Vector :math:`\nabla_{f}(\beta)` is the gradient of the cost function defined above, evaluated at \\(\\beta\\).
```
done

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+        python -m venv benchopt_tutorial_1
+        source benchopt_tutorial_1/bin/activate
+
+Benchopt is referenced in the PyPI index, which means it can be installed using ``pip`` by running in a terminal
```
agreed, done

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,312 @@
+
+.. code-block:: bash 
+
+    pip install benchopt
```
very nice, thanks. Changing all terminal commands blocks to prompt blocks.

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+    g(\beta) = \frac{1}{2} \|y - X\beta \|^2
+
+and the solver implement is GD with \\(\\nabla g(\\beta) = -X^Ty + X^TX\\beta \\) the gradient of \\(g\\) at \\(\\beta \\).
```
```suggestion
and the only implemented solver is GD with \\(\\nabla g(\\beta) = -X^Ty + X^TX\\beta \\) the gradient of \\(g\\) at \\(\\beta \\).
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    :width: 600
+    :alt: A screenshot of the console during the benchopt run command
+
+You will see something similar to this in your terminal
```
add space above if possible

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+Once the benchmark has been run, a window should open in your default navigator.
+This provides a visualization of the results of the run, which is useful to immediately observe, comment and share the results.
+After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset.
```
```suggestion
After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated datasets.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+Once the benchmark has been run, a window should open in your default navigator.
+This provides a visualization of the results of the run, which is useful to immediately observe, comment and share the results.
+After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset.
+The convergence plots can be made log-log for easier reading.
```
```suggestion
The convergence plot scales can be changed for easier reading.
```
but we can also remove this sentence (not essential to understand how to setup a benchmark)

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset.
+The convergence plots can be made log-log for easier reading.
+Also, depending on your computer, the runtime is so low that the convergence plot with respect to time is not reliable.
+We advise to look at the results in terms of iteration (scrolling menu on the bottom left of the webpage).
```
We shouldn't advise that

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+This provides a visualization of the results of the run, which is useful to immediately observe, comment and share the results.
+After running the template benchmark, we can observe the convergence plots of GD with the two different stepsize choices, for two different simulated dataset.
+The convergence plots can be made log-log for easier reading.
+Also, depending on your computer, the runtime is so low that the convergence plot with respect to time is not reliable.
+We advise to look at the results in terms of iteration (scrolling menu on the bottom left of the webpage).
```
```suggestion
In this specific toy example, the runtime is so low that the convergence plot with respect to time may not be reliable: you can also look at the results in terms of iterations, by scrolling the menu on the bottom left of the webpage.
```

cohenjer replied:
OK let's see how things change when modifying the example, committing the suggestion for now

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+Also, depending on your computer, the runtime is so low that the convergence plot with respect to time is not reliable.
+We advise to look at the results in terms of iteration (scrolling menu on the bottom left of the webpage).
+Feel free to play around with the plotting options here!
+Note that the dataset comes with two values for an option whiten, let us ignore this detail in the tutorial.
```
Can you point to a place where this is explained?
```suggestion
Note that the objective comes with two possible values for a  `whiten` parameter. 
Let us ignore this detail in the tutorial.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+The template benchmark we are using at the moment is not exactly encoding the information we need for the ridge regression problem.
+To properly modify the benchmark, first we need to dive deeper into how benchmarks work.
+To follow through this section, it is advised to open an editor (like vscode) in the root of the template benchmark to easily navigate between the files and folders.
```
```suggestion
To follow through this section, it is advised to open an editor (like vscode) at the root of the template benchmark to easily navigate between the files and folders.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    ├── datasets          
+    │   └── simulated.py
+    ├── solvers          
+    │   └── python-gd.py
```
can you rename it to gd here and in the template? simpler for a template IMO

cohenjer replied:
Not sure what you mean; this is the structure of the template benchmark, so I need to match exactly the naming from that benchmark.

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+To properly modify the benchmark, first we need to dive deeper into how benchmarks work.
+To follow through this section, it is advised to open an editor (like vscode) in the root of the template benchmark to easily navigate between the files and folders.
+
+Here is the architecture of our template benchmark:
```
I'm in favor of not displaying the files that we don't speak about : clean template, test config, gitignore. It makes it easier for the reader to focus on the key concepts.

cohenjer replied:
Ok sure

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+The three most important files are
+
+    - ``objective.py``: it contains the information about the cost function we want to minimize. In other words, it defines the formal problem we are interested in.
+    - ``python-gd.py``: it contains the information and code for the gradient descent solver, dedicated to the problem at hand.
```
```suggestion
    - ``solvers/python-gd.py``: it contains the information and code for the gradient descent solver, dedicated to the problem at hand.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+    - ``objective.py``: it contains the information about the cost function we want to minimize. In other words, it defines the formal problem we are interested in.
+    - ``python-gd.py``: it contains the information and code for the gradient descent solver, dedicated to the problem at hand.
+    - ``simulated.py``: it contains the information about the dataset, i.e. the values of \\(y \\) and \\(X \\) used to test the algorithms. All benchmark in fact must have a ``simulated.py`` file which is used for testing by Benchopt.
```
```suggestion
    - ``datasets/simulated.py``: it contains the information about the dataset, i.e. the values of \\(y \\) and \\(X \\) used to test the algorithms. All benchmark in fact must have a ``simulated.py`` file which is used for testing by Benchopt.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+            rng = np.random.RandomState(self.random_state)
+            X = rng.randn(self.n_samples, self.n_features)
+            y = rng.randn(self.n_samples)
+            return dict(X=X, y=y),
```
```suggestion
            return dict(X=X, y=y)
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+Indeed, solvers, dataset and objectives need to exchange information.
+Typically, the solver needs to know the parameters used for the loss, in our case the value of the regularization parameter.
+The objective needs to know the values of \\( X\\) and \\( y\\) defined in the dataset.
+This part of the benchmark can rapidly give you a headache if you are not familiar with the structure of the benchmark, but the figure above should be a good reference point.
```
I'd avoid "headache"

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+This part of the benchmark can rapidly give you a headache if you are not familiar with the structure of the benchmark, but the figure above should be a good reference point.
+When a method from a class feeds a method in another class, it returns a dictionary (such as ``get_data`` we just discussed), otherwise it simply updates the class attributes.
+
+.. The previous figure may be updated to detail these communication methods.
```
rm

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+Therefore we need to implement the following modifications:
+
+    - we should add the regularization term \\( +\\lambda \\|\\beta \\|^2 \\) to the loss in ``objective.py``, and values for the regularization parameter.
+    - we should modify the computed gradient, knowing that \\( \\nabla f(\\beta) = \\nabla g(\\beta) + 2\\lambda\\beta \\).
```
we should also change the name of the benchmark

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    - we should add the regularization term \\( +\\lambda \\|\\beta \\|^2 \\) to the loss in ``objective.py``, and values for the regularization parameter.
+    - we should modify the computed gradient, knowing that \\( \\nabla f(\\beta) = \\nabla g(\\beta) + 2\\lambda\\beta \\).
+
+We will not modify anything in the dataset since the inputs \\(X,y \\) of the regression and ridge regression are essentially the same.
```
```suggestion
We will not modify anything in the dataset since the inputs \\(X,y \\) of the regression and ridge regression are the same.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+Let's start with the ``objective.py`` file.
+The regularization parameter values are part of the formal definition of the problem, so we can define them as attributes of the ``Objective`` class.
+The ``whiten_y`` parameter is already implemented, so we can simply add a ``reg`` parameter in the ``parameters`` dictionary.
```
we can also **replace** the whiten parameter, no?

cohenjer replied:
Ok I removed the True option for simplicity

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+            'reg': [1e1, 1e2]
+        }
+
+This piece of code says that \\( \\lambda\\) should take two values \\( 10\\) or \\( 100\\) across the benchmark.
```
```suggestion
This piece of code says that \\( \\lambda\\) should take two values, \\( 10\\) or \\( 100\\), in the benchmark.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+            return dict(
+                value=.5 * diff.dot(diff) + l2reg,
+                ols=.5 * diff.dot(diff),
+                penalty=l2reg
```
```suggestion
                penalty=l2reg,
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+We have done several modifications here:
+
+    - The ``l2reg`` variable computes the regularization term. It is added to the OLS term in the ``value`` field of the output dictionary. This ``value`` field is the main loss of the benchmark, used by all algorithms to track convergence.
```
add that it MUST be named `value`

cohenjer replied:
Done

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+We have done several modifications here:
+
+    - The ``l2reg`` variable computes the regularization term. It is added to the OLS term in the ``value`` field of the output dictionary. This ``value`` field is the main loss of the benchmark, used by all algorithms to track convergence.
+    - Additional metrics are computes, namely ``ols`` and ``penalty``. Benchmark will compute these metrics alongside the loss function, and we will be able to look at them in the resulting plots.
```
```suggestion
    - Additional metrics are computed, namely ``ols`` and ``penalty``. Benchmark will compute these metrics alongside the loss function, and we will be able to look at them in the resulting plots.
```

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+That's it for the ``objective.py`` file!
+We can now modify the solver.
+Instead of modifying directly ``python-gd.py``, let's create a new solver, ``python-gd-ridge.py``.
```
As said above, I'd modify this solver instead, keeping in mind that it's for beginners.

cohenjer replied:
Done, only one solver and we use the same file

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+.. code-block:: python
+
+    def run(self, n_iter):
+        L = np.linalg.norm(self.X, ord=2) ** 2
```
this is missing `+ self.reg` hence the non convergence in the plots

cohenjer replied:
Good catch, thanks

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    def set_objective(self, X, y, reg):
+        self.X, self.y, self.reg = X, y, reg
+
+As a final step, we can change the name of our new algorithm to ``"GD-ridge"`` in the class attribute ``name``, to differentiate it from ``"GD"`` in the visualization of the results.
```
I'd just say we keep the name (since we modify the existing one)

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+.. image:: screenshot_mytemplate_ridgeloss.png
+    :width: 800
+
+We may observe that the GD-ridge with ``stepsize=1`` reaches a much lower cost value that other methods, in particular GD without the ridge penalisation term.
+This is expected since GD does not minimizes \\(f(\\beta) \\) but rather  \\( g(\\beta) \\).
+Interestingly, using ``stepsize=1.99`` leads to large errors.
+This is caused by the scaling of the stepsize (using the Lipschitz constant) not being updated in our GD-ridge algorithm, and the stepsize is therefore not safe.
```
Too complex and it gives me (my 2 cents) the impression that benchopt is buggy.

I strongly prefer a selling picture where everything is correct and works fine

cohenjer replied:
I will improve on that, stick to ridge only as suggested.

mathurinm commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+This is caused by the scaling of the stepsize (using the Lipschitz constant) not being updated in our GD-ridge algorithm, and the stepsize is therefore not safe.
+
+One of the interesting features of Benchopt is its ability to easily compute and show several metrics over the run.
+We have computed the OLS loss \\(g(\\beta) \\) alongside the iterations, and we can observe its values by changing the ``Objective_column`` field to ``ols``
```
missing concluding dots here and line below

mathurinm reviewed (commented):
Again very nice work @cohenjer 
I'm in favor of some simplifications but overall super happy with the result !

mathurinm commented:
Also, if it's possible to only display the title of the tuto here but not the subsections... 
![image](https://github.com/benchopt/benchopt/assets/8993218/6ee7e98d-a390-47cd-b577-51c4c58f95eb)

cohenjer commented:
Thanks @mathurinm for the in-depth comments! I'm on leave for the coming weeks, will handle that changes when I get back late August.

Badr-MOUFAD commented:
@cohenjer, thanks for this PR!
To let you know, don't worry about the merge conflict, I can handle them afterward.
You can focus on applying @mathurinm remarks.

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    f(\beta) = \frac{1}{2} {\|y - X\beta\|}^2 + \lambda{\|\beta\|}^2
+
+where
+    - \\(y\\) is a vector with \\(m\\) elements, typically some observations
```
I guess that depends on which part of ML you are familiar with :) Maybe we stick to m and n for now, and we can change later if other tutorials/examples make use of a different notation?

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+
+    f(\beta) = \frac{1}{2} {\|y - X\beta\|}^2 + \lambda{\|\beta\|}^2
+
+where
```
This is due to rst automatic formatting for text in front of bullet lists. Should be removed

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+    - \\(y\\) is a vector with \\(m\\) elements, typically some observations
+    - \\(X\\) is a matrix with \\(m \\times n\\) elements, typically the regressors
+    - :math:`\beta` is a vector with \\(n\\) elements that we obtain by regressing \\(y\\) on \\(X\\).
+    - \\(\\|x\\|\\) is the euclidean norm of \\(x\\).
```
done

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+where :math:`\eta` is the learning rate or stepsize, and controls how much the iterate is moved alongside the descent direction.
+The vector :math:`\nabla f(\beta)` is the gradient of the cost function defined above, evaluated at \\(\\beta\\).
+
+Even though GD is a simple algorithm, when applied to our ridge regression problem, we should be mindful of the two hyperparameters: the regularization parameter \\(\\lambda\\), and the gradient stepsize \\( \\eta\\).
```
I tried something, let me know if this works better

cohenjer commented on doc/tutorials/tutorial_understanding_building_benchmark.rst:
```diff
@@ -0,0 +1,370 @@
+.. image:: screenshot_mytemplate_ridgeloss.png
+    :width: 800
+
+We can now dive into the framework and see how to leverage it to easily create our benchmark!
```
With the figure environment, the space between the text and the figure should be more adequate

cohenjer commented:
I updated the benchmark, everything should be good now. Changes:

    - Simplification of the solver, only GD for ridge now
    - Reworked the conclusion/observations
    - Updated all links to match new documentation
    - Small improvements to the text
    
One big issue for me is that the tutorial is broken until the release of 1.5 because template_benchmark is broken, see [this issue](https://github.com/benchopt/template_benchmark/issues/29). I added something in bold in the text for now, we can remove it if we are sure that the tutorial will work with 1.5 release.

tomMoral commented:
> One big issue for me is that the tutorial is broken until the release of 1.5 because template_benchmark is broken, see [this issue](https://github.com/benchopt/template_benchmark/issues/29). I added something in bold in the text for now, we can remove it if we are sure that the tutorial will work with 1.5 release.


I will do a release tomorrow morning so we can test this properly before merging. I was planning to try and get the dev version to `1.5.0.devX` but it seems complicated with the current version of `setuptools_scm` so let's do something simple and ship all the features that are already  in.

tomMoral commented:
It seems that the scheme we consider is not by default in `setuptools_scm` but there is a feature in the upcoming release that will allow to do this: https://github.com/pypa/setuptools_scm#customizing-version-scheme-with-pyprojecttoml
Let's keep this simple for now and release `1.5` now.
For next time, let's consider putting the next micro version as `min_benchopt_version` instead of increasing the minor.

tomMoral commented:
A link to the rendered version: https://output.circle-artifacts.com/output/job/f7905508-fb92-4f8b-81f1-08a95f5a2468/artifacts/0/dev/tutorials/tutorial_understanding_building_benchmark.html

We need to put it with the other tutorials so it is properly linked.

tomMoral reviewed (commented):
Update for 1.5

cohenjer commented:
Thanks @tomMoral, I updated the tutorial according to your recommendations. When 1.5 is out we can merge this unless you want more modifications of the contents, which I would be happy to implement.

tomMoral reviewed (approved):
These are comments I made when I started reviewing this, I don't have the time to do a full review right now so I would say we should ship this as this is already super nice!

Feel free to mer @mathurinm

mathurinm commented:
Thanks for the great work @cohenjer, we can always improve later if some parts need tweaking!
